### PR TITLE
Listener: Filter chain manager refactor cont

### PR DESCRIPTION
**Pull request by silentdai**
Description:
This PR is a follow up of https://github.com/envoyproxy/envoy/pull/7152
The background is FilterChainDiscoveryService(does not exist yet) will update filterchain independently from LDS. 

What this PR does:
1. Moving more factory logic from listener to filter chain manager.
2. Recover the exception message where binded address is lost in #7152 
4.  A scaffold of FilterChainManagerImplTest but not utilized fully.
As this PR is not that trivial, the tests are not moved test from `ListenerTest`. Will do in following PRs.

Risk Level: Mid
Testing: existing Listener test.

### CODE REVIEW

File: source/server/listener_manager_impl.cc [hunk 1]
```diff
@@ -267,120 +270,46 @@ ListenerImpl::ListenerImpl(const envoy::api::v2::Listener& config, const std::st
...
+      parent_.server_.random(), parent_.server_.stats(), parent_.server_.singletonManager(),
+      parent_.server_.threadLocal(), parent_.server_.messageValidationVisitor(),
+      parent_.server_.api());
+  factory_context.setInitManager(initManager());
+  ListenerFilterChainFactoryBuilder builder(*this, factory_context);
+  filter_chain_manager_.addFilterChain(config.filter_chains(), builder);
   // Convert both destination and source IP CIDRs to tries for faster lookups.
   filter_chain_manager_.finishFilterChain();
 
+  bool need_tls_inspector =
```
Comment (rgs1) [inline source/server/listener_manager_impl.cc]: nit: this can be `const bool` now

File: test/server/filter_chain_manager_impl_test.cc [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+using testing::ReturnRef;
+using testing::Throw;
+
+namespace Envoy {
+namespace Server {
+class FilterChainManagerImplTest : public testing::Test {
+public:
+  void SetUp() override {
+    local_address_.reset(new Network::Address::Ipv4Instance("127.0.0.1", 1234));
+    remote_address_.reset(new Network::Address::Ipv4Instance("127.0.0.1", 1234));
```
Comment (rgs1) [inline test/server/filter_chain_manager_impl_test.cc]: nit: `make_shared<>()`

File: test/server/filter_chain_manager_impl_test.cc [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+  std::vector<std::shared_ptr<Network::MockConnectionSocket>> sockets_;
+
+  // Test target
+  FilterChainManagerImpl filter_chain_manager_{
+      std::make_shared<Network::Address::Ipv4Instance>("127.0.0.1", 1234),
+      ProtobufMessage::getNullValidationVisitor()};
+};
+
+TEST_F(FilterChainManagerImplTest, FilterChainMatchNothing) {
+  auto filter_chain =
```
Comment (rgs1) [inline test/server/filter_chain_manager_impl_test.cc]: nit: `const auto*` ?

File: source/server/listener_manager_impl.cc (hunk 1 above)
Comment (silentdai) [reply source/server/listener_manager_impl.cc]: Good catch!

File: test/server/filter_chain_manager_impl_test.cc (hunk 2 above)
Comment (silentdai) [reply test/server/filter_chain_manager_impl_test.cc]: Will fix. I don't like explicit `new` either :)

**Comment by silentdai** [comment]
@dnoe Could you take a look? It's ready to be reviewed.
Also CC @htuch

**Comment by silentdai** [comment]
@dnoe gentle ping

File: source/server/filter_chain_manager_impl.cc [hunk 4]
```diff
@@ -23,19 +32,83 @@ bool FilterChainManagerImpl::isWildcardServerName(const std::string& name) {
...
-    const Protobuf::RepeatedField<Protobuf::uint32>& source_ports,
-    Network::TransportSocketFactoryPtr&& transport_socket_factory,
-    std::vector<Network::FilterFactoryCb> filters_factory) {
-  const auto filter_chain = std::make_shared<FilterChainImpl>(std::move(transport_socket_factory),
-                                                              std::move(filters_factory));
-  addFilterChainForDestinationPorts(destination_ports_map_, destination_port, destination_ips,
-                                    server_names, transport_protocol, application_protocols,
-                                    source_type, source_ips, source_ports, filter_chain);
+    absl::Span<const ::envoy::api::v2::listener::FilterChain* const> filter_chain_span,
+    FilterChainFactoryBuilder& b) {
```
Comment (PiotrSikora) [inline source/server/filter_chain_manager_impl.cc]: Please use proper names, `b` is not descriptive and it's impossible to `grep` for.

File: source/server/filter_chain_manager_impl.cc [hunk 5]
```diff
@@ -23,19 +32,83 @@ bool FilterChainManagerImpl::isWildcardServerName(const std::string& name) {
...
+                                       address_->asString()));
+    }
+    if (filter_chains.find(filter_chain_match) != filter_chains.end()) {
+      throw EnvoyException(fmt::format("error adding listener '{}': multiple filter chains with "
+                                       "the same matching rules are defined",
+                                       address_->asString()));
+    }
+    filter_chains.insert(filter_chain_match);
+
+    // If the cluster doesn't have transport socket configured, then use the default "raw_buffer"
```
Comment (PiotrSikora) [inline source/server/filter_chain_manager_impl.cc]: This is already in `ListenerFilterChainFactoryBuilder::buildFilterChain()`, I don't think you need it here.

File: source/server/filter_chain_manager_impl.cc [hunk 6]
```diff
@@ -23,19 +32,83 @@ bool FilterChainManagerImpl::isWildcardServerName(const std::string& name) {
...
+      if (filter_chain->has_tls_context()) {
+        transport_socket.set_name(Extensions::TransportSockets::TransportSocketNames::get().Tls);
+        MessageUtil::jsonConvert(filter_chain->tls_context(), *transport_socket.mutable_config());
+      } else {
+        transport_socket.set_name(
+            Extensions::TransportSockets::TransportSocketNames::get().RawBuffer);
+      }
+    }
+
+    auto& config_factory = Config::Utility::getAndCheckFactory<
```
Comment (PiotrSikora) [inline source/server/filter_chain_manager_impl.cc]: Same here (already done in `ListenerFilterChainFactoryBuilder::buildFilterChain()`).

File: source/server/filter_chain_manager_impl.cc [hunk 7]
```diff
@@ -23,19 +32,83 @@ bool FilterChainManagerImpl::isWildcardServerName(const std::string& name) {
...
+      if (server_name.find('*') != std::string::npos &&
+          !FilterChainManagerImpl::isWildcardServerName(server_name)) {
+        throw EnvoyException(
+            fmt::format("error adding listener '{}': partial wildcards are not supported in "
+                        "\"server_names\"",
+                        address_->asString()));
+      }
+    }
+
+    std::vector<std::string> source_ips;
```
Comment (PiotrSikora) [inline source/server/filter_chain_manager_impl.cc]: Nit: could you move this right after  `destination_ips`, since this is doing the same validation (yes, I'm aware that this is the order from the old code).

File: source/server/filter_chain_manager_impl.cc [hunk 8]
```diff
@@ -23,19 +32,83 @@ bool FilterChainManagerImpl::isWildcardServerName(const std::string& name) {
...
+    source_ips.reserve(filter_chain_match.source_prefix_ranges().size());
+    for (const auto& source_ip : filter_chain_match.source_prefix_ranges()) {
+      const auto& cidr_range = Network::Address::CidrRange::create(source_ip);
+      source_ips.push_back(cidr_range.asString());
+    }
+
+    std::vector<std::string> application_protocols(
+        filter_chain_match.application_protocols().begin(),
+        filter_chain_match.application_protocols().end());
+    addFilterChainForDestinationPorts(
```
Comment (PiotrSikora) [inline source/server/filter_chain_manager_impl.cc]: Nit: please add an empty line between `application_protocols` and `addFilterChainForDestinationPorts`.

File: source/server/filter_chain_manager_impl.cc [hunk 9]
```diff
@@ -23,19 +32,83 @@ bool FilterChainManagerImpl::isWildcardServerName(const std::string& name) {
...
+
+    // Validate IP addresses.
+    std::vector<std::string> destination_ips;
+    destination_ips.reserve(filter_chain_match.prefix_ranges().size());
+    for (const auto& destination_ip : filter_chain_match.prefix_ranges()) {
+      const auto& cidr_range = Network::Address::CidrRange::create(destination_ip);
+      destination_ips.push_back(cidr_range.asString());
+    }
+
+    std::vector<std::string> server_names(filter_chain_match.server_names().begin(),
```
Comment (PiotrSikora) [inline source/server/filter_chain_manager_impl.cc]: Since filter chain creation is now happening in `ListenerFilterChainFactoryBuilder::buildFilterChain()`, we don't need to create `server_names` here, instead, you can just loop over `filter_chain_match.server_names()` in the code below.

File: source/server/listener_manager_impl.cc [hunk 10]
```diff
@@ -853,5 +779,41 @@ void ListenerManagerImpl::stopWorkers() {
...
+          Extensions::TransportSockets::TransportSocketNames::get().RawBuffer);
+    }
+  }
+
+  auto& config_factory = Config::Utility::getAndCheckFactory<
+      Server::Configuration::DownstreamTransportSocketConfigFactory>(transport_socket.name());
+  ProtobufTypes::MessagePtr message = Config::Utility::translateToFactoryConfig(
+      transport_socket, parent_.messageValidationVisitor(), config_factory);
+
+  // TODO: use span to avoid create vector from repeated fields
```
Comment (PiotrSikora) [inline source/server/listener_manager_impl.cc]: Nit: `TODO(silentdai): ...`

File: test/server/filter_chain_manager_impl_test.cc [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+using testing::ReturnRef;
+using testing::Throw;
+
+namespace Envoy {
+namespace Server {
+
+class MockFilterChainFactoryBuilder : public FilterChainFactoryBuilder {
+  std::unique_ptr<Network::FilterChain>
+  buildFilterChain(const ::envoy::api::v2::listener::FilterChain&) const override {
+    // A place holder to be found
```
Comment (PiotrSikora) [inline test/server/filter_chain_manager_impl_test.cc]: Hm? Could you elaborate on this and/or add a `TODO`?

File: source/server/filter_chain_manager_impl.cc (hunk 4 above)
Comment (silentdai) [reply source/server/filter_chain_manager_impl.cc]: will fix

File: source/server/listener_manager_impl.cc (hunk 10 above)
Comment (silentdai) [reply source/server/listener_manager_impl.cc]: will do

File: test/server/filter_chain_manager_impl_test.cc (hunk 11 above)
Comment (silentdai) [reply test/server/filter_chain_manager_impl_test.cc]: Will add comment:
The test doesn't use the value of the returned ptr but it cannot be nullptr.

File: source/server/filter_chain_manager_impl.cc (hunk 5 above)
Comment (silentdai) [reply source/server/filter_chain_manager_impl.cc]: great catch!

File: source/server/filter_chain_manager_impl.cc (hunk 7 above)
Comment (silentdai) [reply source/server/filter_chain_manager_impl.cc]: done

File: source/server/filter_chain_manager_impl.cc (hunk 9 above)
Comment (silentdai) [reply source/server/filter_chain_manager_impl.cc]: Probably not. 
1. buildFilterChain need a copy of server names.
2. addFilterChainForDestinationPorts currently need a vector as well
3. for loop may not need a vector but 2) need.
I mark 2 as TODO but yet the vector is needed for now

File: source/server/listener_manager_impl.cc [hunk 12]
```diff
@@ -267,120 +270,43 @@ ListenerImpl::ListenerImpl(const envoy::api::v2::Listener& config, const std::st
...
+  const bool need_tls_inspector =
+      std::any_of(
+          config.filter_chains().begin(), config.filter_chains().end(),
+          [](const auto& filter_chain) {
+            const auto& matcher = filter_chain.filter_chain_match();
+            return matcher.transport_protocol() == "tls" ||
+                   (matcher.transport_protocol().empty() &&
+                    (!matcher.server_names().empty() || !matcher.application_protocols().empty()));
+          }) &&
+      std::all_of(config.listener_filters().begin(), config.listener_filters().end(),
```
Comment (PiotrSikora) [inline source/server/listener_manager_impl.cc]: `!std::any_of(... == ...)` might be slightly faster and easier to follow.

File: source/server/filter_chain_manager_impl.cc (hunk 9 above)
Comment (PiotrSikora) [reply source/server/filter_chain_manager_impl.cc]: Ah, right, I've missed `addFilterChainForDestinationPorts()`.

File: test/server/filter_chain_manager_impl_test.cc [hunk 13]
```diff
@@ -0,0 +1,146 @@
...
+using testing::Throw;
+
+namespace Envoy {
+namespace Server {
+
+class MockFilterChainFactoryBuilder : public FilterChainFactoryBuilder {
+
+  std::unique_ptr<Network::FilterChain>
+  buildFilterChain(const ::envoy::api::v2::listener::FilterChain&) const override {
+    // Wont dereference but require not nullptr.
```
Comment (PiotrSikora) [inline test/server/filter_chain_manager_impl_test.cc]: Nit: `s/Wont/Won't/` `s/require/requires/`

File: test/server/filter_chain_manager_impl_test.cc [hunk 14]
```diff
@@ -0,0 +1,146 @@
...
+public:
+  void SetUp() override {
+    local_address_ = std::make_shared<Network::Address::Ipv4Instance>("127.0.0.1", 1234);
+    remote_address_ = std::make_shared<Network::Address::Ipv4Instance>("127.0.0.1", 1234);
+    TestUtility::loadFromYaml(
+        TestEnvironment::substitute(filter_chain_yaml, Network::Address::IpVersion::v4),
+        filter_chain_template_);
+  }
+
+  // Helper for test
```
Comment (PiotrSikora) [inline test/server/filter_chain_manager_impl_test.cc]: `findFilterChainHelper()` name should be self-explanatory ;)

File: test/server/filter_chain_manager_impl_test.cc [hunk 15]
```diff
@@ -0,0 +1,146 @@
...
+    return filter_chain_manager_.findFilterChain(*mock_socket);
+  }
+
+  void addSingleFilterChainHelper(const envoy::api::v2::listener::FilterChain& filter_chain) {
+    filter_chain_manager_.addFilterChain(
+        std::vector<const envoy::api::v2::listener::FilterChain*>{&filter_chain},
+        filter_chain_factory_builder_);
+  }
+
+  // Intermedia state
```
Comment (PiotrSikora) [inline test/server/filter_chain_manager_impl_test.cc]: Nit: missing `.` at the end of comments (here and in other places).

File: source/server/filter_chain_manager_impl.h [hunk 16]
```diff
@@ -122,12 +127,13 @@ class FilterChainManagerImpl : public Network::FilterChainManager,
   // Mapping of FilterChain's configured destination ports, IPs, server names, transport protocols
   // and application protocols, using structures defined above.
   DestinationPortsMap destination_ports_map_;
+  Network::Address::InstanceConstSharedPtr address_;
```
Comment (mattklein123) [inline source/server/filter_chain_manager_impl.h]: const

File: source/server/filter_chain_manager_impl.cc [hunk 17]
```diff
@@ -1,6 +1,15 @@
 #include "server/filter_chain_manager_impl.h"
 
 #include "common/common/empty_string.h"
+#include "common/common/fmt.h"
+#include "common/config/utility.h"
+#include "common/protobuf/utility.h"
+
+#include "server/configuration_impl.h"
+
+#include "extensions/filters/listener/well_known_names.h"
```
Comment (mattklein123) [inline source/server/filter_chain_manager_impl.cc]: Why are these needed? In general I would like to get rid of these headers (tracked in an issue I opened), so it would be nice to not include them here.

File: source/server/filter_chain_manager_impl.h [hunk 18]
```diff
@@ -13,26 +15,29 @@
 namespace Envoy {
 namespace Server {
 
+class FilterChainFactoryBuilder {
+public:
+  virtual ~FilterChainFactoryBuilder() = default;
+  virtual std::unique_ptr<Network::FilterChain>
+  buildFilterChain(const ::envoy::api::v2::listener::FilterChain& filter_chain) const = 0;
```
Comment (mattklein123) [inline source/server/filter_chain_manager_impl.h]: nit: s/= 0/PURE

File: source/server/listener_manager_impl.h [hunk 19]
```diff
@@ -19,9 +19,19 @@
 #include "server/filter_chain_manager_impl.h"
 #include "server/lds_api.h"
 
+#include "extensions/filters/listener/well_known_names.h"
```
Comment (mattklein123) [inline source/server/listener_manager_impl.h]: Why are these needed?

File: source/server/listener_manager_impl.cc [hunk 20]
```diff
@@ -267,120 +270,43 @@ ListenerImpl::ListenerImpl(const envoy::api::v2::Listener& config, const std::st
...
+  const bool need_tls_inspector =
+      std::any_of(
+          config.filter_chains().begin(), config.filter_chains().end(),
+          [](const auto& filter_chain) {
+            const auto& matcher = filter_chain.filter_chain_match();
+            return matcher.transport_protocol() == "tls" ||
+                   (matcher.transport_protocol().empty() &&
+                    (!matcher.server_names().empty() || !matcher.application_protocols().empty()));
+          }) &&
+      not std::any_of(config.listener_filters().begin(), config.listener_filters().end(),
```
Comment (mattklein123) [inline source/server/listener_manager_impl.cc]: TIL that `not` is a valid keyword in C++. Mind blown.

File: source/server/filter_chain_manager_impl.h [hunk 21]
```diff
@@ -13,26 +15,29 @@
...
-                 const std::vector<std::string>& server_names,
-                 const std::string& transport_protocol,
-                 const std::vector<std::string>& application_protocols,
-                 const envoy::api::v2::listener::FilterChainMatch_ConnectionSourceType source_type,
-                 const std::vector<std::string>& source_ips,
-                 const Protobuf::RepeatedField<Protobuf::uint32>& source_ports,
-                 Network::TransportSocketFactoryPtr&& transport_socket_factory,
-                 std::vector<Network::FilterFactoryCb> filters_factory);
-  void finishFilterChain() { convertIPsToTries(); }
+  addFilterChain(absl::Span<const ::envoy::api::v2::listener::FilterChain* const> filter_chain_span,
```
Comment (mattklein123) [inline source/server/filter_chain_manager_impl.h]: I'm not sure it's really worth converting this to a span? Can we just pass the relevant proto container directly via const reference?

File: source/server/filter_chain_manager_impl.h [hunk 22]
```diff
@@ -13,26 +15,29 @@
...
+  buildFilterChain(const ::envoy::api::v2::listener::FilterChain& filter_chain) const = 0;
+};
+
 /**
  * Implementation of FilterChainManager.
  */
 class FilterChainManagerImpl : public Network::FilterChainManager,
                                Logger::Loggable<Logger::Id::config> {
 public:
+  explicit FilterChainManagerImpl(Network::Address::InstanceConstSharedPtr address)
```
[the remark this answers is not in the document]
Comment (silentdai) [reply source/server/filter_chain_manager_impl.h]: great catch! will fix

File: source/server/filter_chain_manager_impl.cc (hunk 17 above)
Comment (silentdai) [reply source/server/filter_chain_manager_impl.cc]: There is a tricky flow that `tls_inspector` is implicitly added when filterchain matcher has the rule alpn. Not sure if it is needed in this PR. Will double check

File: source/server/filter_chain_manager_impl.h (hunk 18 above)
Comment (silentdai) [reply source/server/filter_chain_manager_impl.h]: oops

File: source/server/listener_manager_impl.h (hunk 19 above)
Comment (silentdai) [reply source/server/listener_manager_impl.h]: The intention is to add tls_inspector. Will double check

File: source/server/filter_chain_manager_impl.h (hunk 21 above)
Comment (silentdai) [reply source/server/filter_chain_manager_impl.h]: The idea is to provide as few information as possible, and my limited knowledge on protobuf structure :)

By `fewer message` my point is using span so we can use "vector of  FilterChain" in test and "a proto container" in prod. They should be functional equivalent from the perspective of this method. So is the performance from the declaration of `absl::Span`

Comment (mattklein123) [reply source/server/filter_chain_manager_impl.h]: OK that's fine.

File: source/server/filter_chain_manager_impl.cc (hunk 17 above)
Comment (silentdai) [reply source/server/filter_chain_manager_impl.cc]: Removed header and bazel deps

File: source/server/listener_manager_impl.h (hunk 19 above)
Comment (silentdai) [reply source/server/listener_manager_impl.h]: fixed

**Comment by silentdai** [comment]
@mattklein123 I resolve all the change requests
PTAL when you are available

**Comment by silentdai** [comment]
gentle ping @mattklein123

**Comment by silentdai** [comment]
Thank you for allowing me leave some TODOs in this PR! I will post the follow up later today.
